File: LeetCode/98.Validate_Binary_Search_Tree/Mario/98.Validate_Binary_Search_Tree.py
#!/usr/bin/env python3


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.helper(root, float("-inf"), float("inf"))

    def helper(self, root, min, max):
        if not root:
            return True
        if min >= root.val or max <= root.val:
            return False
        res = self.helper(root.left, min, root.val) and self.helper(
            root.right, root.val, max
        )
        return res


# inorder traverse solution
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        self.pre = None
        self.res = True

        def inorder(root):
            # leaf or self.res is False
            if not root or not self.res:
                return

            inorder(root.left)
            if self.pre is not None:
                if root.val <= self.pre:
                    self.res = False
            self.pre = root.val
            inorder(root.right)

        inorder(root)

        return self.res

File: LeetCode/98.Validate_Binary_Search_Tree/Mario/98.Validate_Binary_Search_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return dfs(root, null, null);
    }

    private boolean dfs(TreeNode node, Integer maxValue, Integer minValue) {
        if (node == null)
            return true;
        if ((minValue == null || node.val > minValue) && (maxValue == null || node.val < maxValue)) {
            return dfs(node.left, node.val, minValue) && dfs(node.right, maxValue, node.val);
        }
        return false;
    }
}